winpr/wlog.1
.\" Written by David Fort (contact@hardening-consulting.com)
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH wLog 13 "June 2016" Version "1.0"
.SH NAME
wLog \- WinPR logging facility

.SH DESCRIPTION
wLog is a configurable and flexible logging system used throughout WinPR and
FreeRDP.

The primary concept is to have a hierarchy of loggers that can be be configured
independently.

.SH Appenders

WLog uses different appenders that define where the log output should be written
to. If the application doesn't explicitly configure the appenders the below
described variable WLOG_APPENDER can be used to choose one appender.

The following kind of appenders are available:

.IP Binary
Write the log data into a binary format file.

.IP Console
The console appender writes to the console. Depending of the operating system
the application runs on, the output might be handled differently. For example
on android log print would be used.

.IP File
The file appender writes the textual output to a file.

.IP Udp
This appender sends the logging messages to a pre-defined remote host via UDP.

If no target is set the default one 127.0.0.1:20000 is used. To receive the
log messages one can use netcat. To receive the default target the following
command can be used:
nc -u 127.0.0.1 -p 20000 -l
.IP Syslog
Use syslog for outputting the debug messages.
.IP Journald
This appender outputs messages to journald.

.SH Levels
The WLog are complementary, the higher level always includes the lower ones.
The level list below is top down. Top the highest level.

.IP TRACE 
print everything including packets dumps
.IP DEBUG 
debug messages
.IP INFO
general informations
.IP WARN
warnings
.IP ERROR
errors
.IP FATAL
fatal problems
.IP OFF
completely disable the wlog output

.SH Formats
The format a logger prints has the following possible options:

.IP lv
log level
.IP mn
module name
.IP fl
file name
.IP fn
function
.IP ln
line number
.IP pid
process id
.IP tid
thread id
.IP yr
year
.IP mo
month
.IP dw
day of week
.IP hr
hour
.IP mi
minute
.IP se
second
.IP ml
millisecond
.PP
A maximum of 16 options can be used per format string.

An example that generally sets the WLOG_PREFIX for xfreerdp would look like:
WLOG_PREFIX="pid=%pid:tid=%tid:fn=%fn -" xfreerdp /v:xxx


.SH ENVIRONMENT
.IP WLOG_APPENDER
The kind of appender, the accepted values are: CONSOLE, FILE, BINARY, SYSLOG, JOURNALD or UDP
  
.IP WLOG_PREFIX
configure the prefix used for outputting the message (see Format for more details and examples)

.IP WLOG_LEVEL
the level to output messages for 

.IP WLOG_FILTER
sets a filter for WLog messages. Only the filtered messages are
printed. The format of the filter is a serie of \<logger name\>:\<level\> separated by
comas

example: WLOG_FILTER=core.channel:DEBUG,dummy:TRACE
will display debug messages for the core.channel logger and trace level for the dummy logger

.IP WLOG_FILEAPPENDER_OUTPUT_FILE_PATH
When using the file appender it may contains the output log file's path 

.IP WLOG_FILEAPPENDER_OUTPUT_FILE_NAME
When using the file appender it may contains the output log file's name

.IP WLOG_JOURNALD_ID
When using the systemd journal appender, this variable contains the id used with
the journal (by default the executable's name)

.IP WLOG_UDP_TARGET
target to use for the UDP appender in the format 
.B host:port

.SH BUGS
Please report any bugs using the bug reporting form on the
.B FreeRDP
web site 

.SH "SEE ALSO"
Additional information and the latest version is available
at the web site:
.B http://www.freerdp.com

.SH AUTHOR
David Fort <contact@hardening-consulting.com> wrote this manpage from materials written
by Bernhard Miklautz <bernhard.miklautz@thincast.com>.



ChangeLog
This version optimized to run dfreerdp on Raspberry PI

2013-01-02 Version 1.0.2

FreeRDP 1.0.2 is a maintenance release which contains several bug and stability fixes.

* xfreerdp:
	* new parameter --from-stdin - prompts for unspecified arguments username, password, domain and host.
	* fix compability with x2go
	* fix keyboard state in remote app
	* documentation fixes
	* fixed crash when started with --authonly (#843)

* libfreerdp-core:
	* several memory leaks and double frees were fixed
	* support for FastPath PDUs up to 32767
	* register audio only if plugin is registered
	* load extensions after argument parsing

* libfreerdp-utils:
	* fixed crash when HOME environment variable wasn't set

* xfreerdp-server
	* deadlock fixed
	* accept TLSv1 and SSLv3

* smartcard
	* don't incorrectly set SCARD_STATE_IGNORE

* libfreerdp-codec
	* performance improvement

* libfreerdp-gdi
	* support for PatBlt DPa operation

* plugin
	* ignore CHANNEL_FLAG_SUSPEND/CHANNEL_FLAG_RESUME to prevent possible crash

For a detailed list of changes use "git log 1.0.1..1.0.2"

Known problems:

* If windows input language is set to german pressing the divde key (/) on the keypad 
  results in minus (-) (see issue #811)

2012-02-07	Version 1.0.1

FreeRDP 1.0.1 is a maintenance release to address a certain number of
issues found in 1.0.0. This release also brings corrective measures
to certificate validation which were required for inclusion in Ubuntu.

* Certificate Validation
	* Improved validation logic and robustness
	* Added validation of certificate name against hostname

* Token-based Server Redirection
	* Fixed redirection logic
	* HAProxy load-balancer support

* xfreerdp-server
	* better event handling
	* capture performance improvements

* wfreerdp
	* Fix RemoteFX support
	* Fix mingw64 compilation

* libfreerdp-core:
	* Fix severe TCP sending bug
	* Added server-side Standard RDP security

2012-01-16	Version 1.0.0

License:

FreeRDP 1.0 is the first release of FreeRDP under the Apache License 2.0.
The FreeRDP 1.x series is a rewrite, meaning there is no continuity with
the previous FreeRDP 0.x series which were released under GPLv2.

New Features:

* RemoteFX
	* Both encoder and decoder
	* SSE2 and NEON optimization
* NSCodec
* RemoteApp
	* Working, minor glitches
* Multimedia Redirection
	* ffmpeg support
* Network Level Authentication (NLA)
	* NTLMv2
* Certificate validation
* FIPS-compliant RDP security
* new build system (cmake)
* added official logo and icon

New Architecture:

* libfreerdp-core
	* core protocol
	* highly portable
	* both client and server
* libfreerdp-cache
	* caching operations
* libfreerdp-codec
	* bitmap decompression
	* codec encoding/decoding
* libfreerdp-kbd
	* keyboard mapping
* libfreerdp-channels
	* virtual channel management
	* client and server side support
* libfreerdp-gdi
	* extensively unit tested
	* portable software GDI implementation
* libfreerdp-rail
	* RemoteApp library
* libfreerdp-utils
	* shared utility library

FreeRDP Clients:

* client/X11 (xfreerdp)
	* official client
	* RemoteApp support
	* X11 GDI implementation
* client/DirectFB (dfreerdp)
	* DirectFB support
	* software-based GDI (libfreerdp-gdi)
* client/Windows (wfreerdp)
	* Native Win32 support

FreeRDP Servers (experimental):

* server/X11 (xfreerdp-server)
	* RemoteFX-only
	* no authentication
	* highly experimental
	* keyboard and mouse input supported

Virtual Channels:

* cliprdr (Clipboard Redirection)
* rail (RemoteApp)
* drdynvc (Dynamic Virtual Channels)
	* audin (Audio Input Redirection)
		* alsa support
		* pulse support
	* tsmf (Multimedia Redirection)
		* alsa support
		* pulse support
		* ffmpeg support
* rdpdr (Device Redirection)
	* disk (Disk Redirection)
	* parallel (Parallel Port Redirection)
	* serial (Serial Port Redirection)
	* printer (Printer Redirection)
		* CUPS support
	* smartcard (Smartcard Redirection)
* rdpsnd (Sound Redirection)
	* alsa support
	* pulse support



winpr/libwinpr/utils/wlog/wlog.c
/**
 * WinPR: Windows Portable Runtime
 * WinPR Logger
 *
 * Copyright 2013 Marc-Andre Moreau <marcandre.moreau@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <string.h>

#include <winpr/crt.h>
#include <winpr/print.h>
#include <winpr/debug.h>
#include <winpr/environment.h>

#if defined(ANDROID)
#include <android/log.h>
#include "../log.h"
#endif

#include "wlog.h"


struct _wLogFilter
{
	DWORD Level;
	LPSTR* Names;
	DWORD NameCount;
};
typedef struct _wLogFilter wLogFilter;

/**
 * References for general logging concepts:
 *
 * Short introduction to log4j:
 * http://logging.apache.org/log4j/1.2/manual.html
 *
 * logging - Logging facility for Python:
 * http://docs.python.org/2/library/logging.html
 */

LPCSTR WLOG_LEVELS[7] =
{
	"TRACE",
	"DEBUG",
	"INFO",
	"WARN",
	"ERROR",
	"FATAL",
	"OFF"
};

static DWORD g_FilterCount = 0;
static wLogFilter* g_Filters = NULL;

static LONG WLog_GetFilterLogLevel(wLog* log);
static int WLog_ParseLogLevel(LPCSTR level);
static BOOL WLog_ParseFilter(wLogFilter* filter, LPCSTR name);

static BOOL log_recursion(LPCSTR file, LPCSTR fkt, int line)
{
	char** msg;
	size_t used, i;
	void* bt = winpr_backtrace(20);
#if defined(ANDROID)
	LPCSTR tag = WINPR_TAG("utils.wlog");
#endif

	if (!bt)
		return FALSE;
	msg = winpr_backtrace_symbols(bt, &used);
	if (!msg)
		return FALSE;
#if defined(ANDROID)
	if (__android_log_print(ANDROID_LOG_FATAL, tag, "Recursion detected!!!") < 0)
		return FALSE;
	if (__android_log_print(ANDROID_LOG_FATAL, tag, "Check %s [%s:%d]", fkt, file, line) < 0)
		return FALSE;

	for (i=0; i<used; i++)
		if (__android_log_print(ANDROID_LOG_FATAL, tag, "%d: %s", i, msg[i]) < 0)
			return FALSE;

#else
	if (fprintf(stderr, "[%s]: Recursion detected!\n", fkt) < 0)
		return FALSE;
	if (fprintf(stderr, "[%s]: Check %s:%d\n", fkt, file, line) < 0)
		return FALSE;

	for (i=0; i<used; i++)
		if (fprintf(stderr, "%s: %lu: %s\n", fkt, (unsigned long)i, msg[i]) < 0)
			return FALSE;

#endif

	free(msg);

	winpr_backtrace_free(bt);
	return TRUE;
}

BOOL WLog_Write(wLog* log, wLogMessage* message)
{
	BOOL status;
	wLogAppender* appender;
	appender = WLog_GetLogAppender(log);

	if (!appender)
		return FALSE;

	if (!appender->active)
		if (!WLog_OpenAppender(log))
			return FALSE;

	if (!appender->WriteMessage)
		return FALSE;

	EnterCriticalSection(&appender->lock);

	if (appender->recursive)
		status = log_recursion(message->FileName, message->FunctionName, message->LineNumber);
	else
	{
		appender->recursive = TRUE;
		status = appender->WriteMessage(log, appender, message);
		appender->recursive = FALSE;
	}

	LeaveCriticalSection(&appender->lock);
	return status;
}

BOOL WLog_WriteData(wLog* log, wLogMessage* message)
{
	BOOL status;
	wLogAppender* appender;
	appender = WLog_GetLogAppender(log);

	if (!appender)
		return FALSE;

	if (!appender->active)
		if (!WLog_OpenAppender(log))
			return FALSE;

	if (!appender->WriteDataMessage)
		return FALSE;

	EnterCriticalSection(&appender->lock);

	if (appender->recursive)
		status = log_recursion(message->FileName, message->FunctionName, message->LineNumber);
	else
	{
		appender->recursive = TRUE;
		status = appender->WriteDataMessage(log, appender, message);
		appender->recursive = FALSE;
	}

	LeaveCriticalSection(&appender->lock);
	return status;
}

BOOL WLog_WriteImage(wLog* log, wLogMessage* message)
{
	BOOL status;
	wLogAppender* appender;
	appender = WLog_GetLogAppender(log);

	if (!appender)
		return FALSE;

	if (!appender->active)
		if (!WLog_OpenAppender(log))
			return FALSE;

	if (!appender->WriteImageMessage)
		return FALSE;

	EnterCriticalSection(&appender->lock);

	if (appender->recursive)
		status = log_recursion(message->FileName, message->FunctionName, message->LineNumber);
	else
	{
		appender->recursive = TRUE;
		status = appender->WriteImageMessage(log, appender, message);
		appender->recursive = FALSE;
	}

	LeaveCriticalSection(&appender->lock);
	return status;
}

BOOL WLog_WritePacket(wLog* log, wLogMessage* message)
{
	BOOL status;
	wLogAppender* appender;
	appender = WLog_GetLogAppender(log);

	if (!appender)
		return FALSE;

	if (!appender->active)
		if (!WLog_OpenAppender(log))
			return FALSE;

	if (!appender->WritePacketMessage)
		return FALSE;

	EnterCriticalSection(&appender->lock);

	if (appender->recursive)
		status = log_recursion(message->FileName, message->FunctionName, message->LineNumber);
	else
	{
		appender->recursive = TRUE;
		status = appender->WritePacketMessage(log, appender, message);
		appender->recursive = FALSE;
	}

	LeaveCriticalSection(&appender->lock);
	return status;
}

BOOL WLog_PrintMessageVA(wLog* log, wLogMessage* message, va_list args)
{
	BOOL status = FALSE;

	if (message->Type == WLOG_MESSAGE_TEXT)
	{
		if (!strchr(message->FormatString, '%'))
		{
			message->TextString = (LPSTR) message->FormatString;
			status = WLog_Write(log, message);
		}
		else
		{
			char formattedLogMessage[WLOG_MAX_STRING_SIZE];
			if (wvsnprintfx(formattedLogMessage, WLOG_MAX_STRING_SIZE - 1, message->FormatString, args) < 0)
				return FALSE;
			message->TextString = formattedLogMessage;
			status = WLog_Write(log, message);
		}
	}
	else if (message->Type == WLOG_MESSAGE_DATA)
	{
		message->Data = va_arg(args, void*);
		message->Length = va_arg(args, int);
		status = WLog_WriteData(log, message);
	}
	else if (message->Type == WLOG_MESSAGE_IMAGE)
	{
		message->ImageData = va_arg(args, void*);
		message->ImageWidth = va_arg(args, int);
		message->ImageHeight = va_arg(args, int);
		message->ImageBpp = va_arg(args, int);
		status = WLog_WriteImage(log, message);
	}
	else if (message->Type == WLOG_MESSAGE_PACKET)
	{
		message->PacketData = va_arg(args, void*);
		message->PacketLength = va_arg(args, int);
		message->PacketFlags = va_arg(args, int);
		status = WLog_WritePacket(log, message);
	}

	return status;
}

BOOL WLog_PrintMessage(wLog* log, wLogMessage* message, ...)
{
	BOOL status;
	va_list args;
	va_start(args, message);
	status = WLog_PrintMessageVA(log, message, args);
	va_end(args);
	return status;
}

DWORD WLog_GetLogLevel(wLog* log)
{
	if (log->FilterLevel < 0)
		log->FilterLevel = WLog_GetFilterLogLevel(log);

	if ((log->FilterLevel >= 0) && (log->FilterLevel != WLOG_LEVEL_INHERIT))
		return log->FilterLevel;
	else if (log->Level == WLOG_LEVEL_INHERIT)
		log->Level = WLog_GetLogLevel(log->Parent);
	return log->Level;
}

BOOL WLog_SetStringLogLevel(wLog* log, LPCSTR level)
{
	int lvl;
	if (!log || !level)
		return FALSE;

	lvl = WLog_ParseLogLevel(level);
	if (lvl < 0)
		return FALSE;

	return WLog_SetLogLevel(log, lvl);
}

BOOL WLog_AddStringLogFilters(LPCSTR filter)
{
	DWORD pos;
	DWORD size;
	DWORD count;
	LPSTR p;
	LPSTR filterStr;
	LPSTR cp;
	wLogFilter* tmp;

	if (!filter)
		return FALSE;

	count = 1;
	p = (LPSTR)filter;

	while ((p = strchr(p, ',')) != NULL)
	{
		count++;
		p++;
	}

	pos = g_FilterCount;
	size = g_FilterCount + count;
	tmp = (wLogFilter*) realloc(g_Filters, size * sizeof(wLogFilter));

	if (!tmp)
		return FALSE;

	g_Filters = tmp;

	cp = (LPSTR)_strdup(filter);
	if (!cp)
		return FALSE;

	p = cp;
	filterStr = cp;

	do
	{
		p = strchr(p, ',');
		if (p)
			*p = '\0';

		if (pos < size)
		{
			if (!WLog_ParseFilter(&g_Filters[pos++], filterStr))
			{
				free (cp);
				return FALSE;
			}
		}
		else
			break;

		if (p)
		{
			filterStr = p + 1;
			p++;
		}
	}
	while (p != NULL);

	g_FilterCount = size;
	free (cp);

	return TRUE;
}

BOOL WLog_SetLogLevel(wLog* log, DWORD logLevel)
{
	if (!log)
		return FALSE;

	if ((logLevel > WLOG_OFF) && (logLevel != WLOG_LEVEL_INHERIT))
		logLevel = WLOG_OFF;

	log->Level = logLevel;
	return TRUE;
}

int WLog_ParseLogLevel(LPCSTR level)
{
	int iLevel = -1;

	if (!level)
		return -1;

	if (_stricmp(level, "TRACE") == 0)
		iLevel = WLOG_TRACE;
	else if (_stricmp(level, "DEBUG") == 0)
		iLevel = WLOG_DEBUG;
	else if (_stricmp(level, "INFO") == 0)
		iLevel = WLOG_INFO;
	else if (_stricmp(level, "WARN") == 0)
		iLevel = WLOG_WARN;
	else if (_stricmp(level, "ERROR") == 0)
		iLevel = WLOG_ERROR;
	else if (_stricmp(level, "FATAL") == 0)
		iLevel = WLOG_FATAL;
	else if (_stricmp(level, "OFF") == 0)
		iLevel = WLOG_OFF;

	return iLevel;
}

BOOL WLog_ParseFilter(wLogFilter* filter, LPCSTR name)
{
	char* p;
	char* q;
	int count;
	LPSTR names;
	int iLevel;
	count = 1;

	if(!name)
		return FALSE;

	p = (char*) name;

	if (p)
	{
		while ((p = strchr(p, '.')) != NULL)
		{
			count++;
			p++;
		}
	}

	names = _strdup(name);
	if (!names)
		return FALSE;
	filter->NameCount = count;
	filter->Names = (LPSTR*) calloc((count + 1UL), sizeof(LPSTR));
	if(!filter->Names)
	{
		free(names);
		filter->NameCount = 0;
		return FALSE;
	}
	filter->Names[count] = NULL;
	count = 0;
	p = (char*) names;
	filter->Names[count++] = p;
	q = strrchr(p, ':');

	if (!q)
	{
		free(names);
		free(filter->Names);
		filter->Names = NULL;
		filter->NameCount = 0;
		return FALSE;
	}

	*q = '\0';
	q++;
	iLevel = WLog_ParseLogLevel(q);

	if (iLevel < 0)
	{
		free(names);
		free(filter->Names);
		filter->Names = NULL;
		filter->NameCount = 0;
		return FALSE;
	}

	filter->Level = (DWORD) iLevel;

	while ((p = strchr(p, '.')) != NULL)
	{
		if (count < (int) filter->NameCount)
			filter->Names[count++] = p + 1;
		*p = '\0';
		p++;
	}

	return TRUE;
}

BOOL WLog_ParseFilters(void)
{
	BOOL res;
	char* env;
	DWORD nSize;

	g_Filters = NULL;
	g_FilterCount = 0;

	nSize = GetEnvironmentVariableA("WLOG_FILTER", NULL, 0);

	if (nSize < 1)
		return TRUE;

	env = (LPSTR) malloc(nSize);

	if (!env)
		return FALSE;

	if (!GetEnvironmentVariableA("WLOG_FILTER", env, nSize))
		return FALSE;

	res = WLog_AddStringLogFilters(env);

	free(env);

	return res;
}

LONG WLog_GetFilterLogLevel(wLog* log)
{
	DWORD i, j;
	BOOL match = FALSE;

	if (log->FilterLevel >= 0)
		return log->FilterLevel;

	for (i = 0; i < g_FilterCount; i++)
	{
		for (j = 0; j < g_Filters[i].NameCount; j++)
		{
			if (j >= log->NameCount)
				break;

			if (_stricmp(g_Filters[i].Names[j], "*") == 0)
			{
				match = TRUE;
				break;
			}

			if (_stricmp(g_Filters[i].Names[j], log->Names[j]) != 0)
				break;

			if (j == (log->NameCount - 1))
			{
				match = TRUE;
				break;
			}
		}

		if (match)
			break;
	}

	if (match)
		log->FilterLevel = g_Filters[i].Level;
	else
		log->FilterLevel = WLOG_LEVEL_INHERIT;

	return log->FilterLevel;
}

BOOL WLog_ParseName(wLog* log, LPCSTR name)
{
	char* p;
	int count;
	LPSTR names;
	count = 1;
	p = (char*) name;

	while ((p = strchr(p, '.')) != NULL)
	{
		count++;
		p++;
	}

	names = _strdup(name);
	if (!names)
		return FALSE;
	log->NameCount = count;
	log->Names = (LPSTR*) calloc((count + 1UL), sizeof(LPSTR));
	if(!log->Names)
	{
		free(names);
		return FALSE;
	}
	log->Names[count] = NULL;
	count = 0;
	p = (char*) names;
	log->Names[count++] = p;

	while ((p = strchr(p, '.')) != NULL)
	{
		if (count < (int) log->NameCount)
			log->Names[count++] = p + 1;
		*p = '\0';
		p++;
	}

	return TRUE;
}

wLog* WLog_New(LPCSTR name, wLog* rootLogger)
{
	wLog* log = NULL;
	char* env = NULL;
	DWORD nSize;
	int iLevel;

	log = (wLog*) calloc(1, sizeof(wLog));
	if (!log)
		return NULL;

	log->Name = _strdup(name);

	if (!log->Name)
		goto out_fail;

	if (!WLog_ParseName(log, name))
		goto out_fail;

	log->Parent = rootLogger;
	log->ChildrenCount = 0;
	log->ChildrenSize = 16;
	log->FilterLevel = -1;

	if (!(log->Children = (wLog**) calloc(log->ChildrenSize, sizeof(wLog*))))
		goto out_fail;

	log->Appender = NULL;

	if (rootLogger)
	{
		log->Level = WLOG_LEVEL_INHERIT;
	}
	else
	{
		log->Level = WLOG_INFO;
		nSize = GetEnvironmentVariableA("WLOG_LEVEL", NULL, 0);

		if (nSize)
		{
			env = (LPSTR) malloc(nSize);
			if (!env)
				goto out_fail;

			if (!GetEnvironmentVariableA("WLOG_LEVEL", env, nSize))
			{
				fprintf(stderr, "WLOG_LEVEL environment variable changed in my back !\n");
				free(env);
				goto out_fail;
			}

			iLevel = WLog_ParseLogLevel(env);
			free(env);

			if (iLevel >= 0)
				log->Level = (DWORD) iLevel;
		}
	}

	iLevel = WLog_GetFilterLogLevel(log);

	if ((iLevel >= 0) && (iLevel != WLOG_LEVEL_INHERIT))
		log->Level = (DWORD) iLevel;

	return log;

out_fail:
	free (log->Children);
	free (log->Name);
	free (log);
	return NULL;
}

void WLog_Free(wLog* log)
{
	if (log)
	{
		if (log->Appender)
		{
			WLog_Appender_Free(log, log->Appender);
			log->Appender = NULL;
		}

		free(log->Name);
		free(log->Names[0]);
		free(log->Names);
		free(log->Children);
		free(log);
	}
}

static wLog* g_RootLog = NULL;

wLog* WLog_GetRoot(void)
{
	char* env;
	DWORD nSize;
	DWORD logAppenderType;

	if (!g_RootLog)
	{
		if (!(g_RootLog = WLog_New("", NULL)))
			return NULL;

		g_RootLog->IsRoot = TRUE;
		WLog_ParseFilters();
		logAppenderType = WLOG_APPENDER_CONSOLE;
		nSize = GetEnvironmentVariableA("WLOG_APPENDER", NULL, 0);

		if (nSize)
		{
			env = (LPSTR) malloc(nSize);
			if (!env)
				goto fail;

			if (!GetEnvironmentVariableA("WLOG_APPENDER", env, nSize))
			{
				fprintf(stderr, "WLOG_APPENDER environment variable modified in my back");
				free(env);
				goto fail;
			}

			if (_stricmp(env, "CONSOLE") == 0)
				logAppenderType = WLOG_APPENDER_CONSOLE;
			else if (_stricmp(env, "FILE") == 0)
				logAppenderType = WLOG_APPENDER_FILE;
			else if (_stricmp(env, "BINARY") == 0)
				logAppenderType = WLOG_APPENDER_BINARY;
#ifdef HAVE_SYSLOG_H
			else if (_stricmp(env, "SYSLOG") == 0)
				logAppenderType = WLOG_APPENDER_SYSLOG;
#endif /* HAVE_SYSLOG_H */
#ifdef HAVE_JOURNALD_H
			else if (_stricmp(env, "JOURNALD") == 0)
				logAppenderType = WLOG_APPENDER_JOURNALD;
#endif
			else if (_stricmp(env, "UDP") == 0)
				logAppenderType = WLOG_APPENDER_UDP;

			free(env);
		}

		if (!WLog_SetLogAppenderType(g_RootLog, logAppenderType))
			goto fail;
	}

	return g_RootLog;

fail:
	free(g_RootLog);
	g_RootLog = NULL;
	return NULL;
}

BOOL WLog_AddChild(wLog* parent, wLog* child)
{
	if (parent->ChildrenCount >= parent->ChildrenSize)
	{
		wLog **tmp;
		parent->ChildrenSize *= 2;
		if (!parent->ChildrenSize)
		{
			if (parent->Children)
				free (parent->Children);
			parent->Children = NULL;
		}
		else
		{
			tmp = (wLog**) realloc(parent->Children, sizeof(wLog*) * parent->ChildrenSize);
			if (!tmp)
			{
				if (parent->Children)
					free (parent->Children);
				parent->Children = NULL;
				return FALSE;
			}
			parent->Children = tmp;
		}
	}

	if (!parent->Children)
		return FALSE;

	parent->Children[parent->ChildrenCount++] = child;
	child->Parent = parent;
	return TRUE;
}

wLog* WLog_FindChild(LPCSTR name)
{
	DWORD index;
	wLog* root;
	wLog* child = NULL;
	BOOL found = FALSE;
	root = WLog_GetRoot();

	if (!root)
		return NULL;

	for (index = 0; index < root->ChildrenCount; index++)
	{
		child = root->Children[index];

		if (strcmp(child->Name, name) == 0)
		{
			found = TRUE;
			break;
		}
	}

	return (found) ? child : NULL;
}

wLog* WLog_Get(LPCSTR name)
{
	wLog* log;
	if (!(log = WLog_FindChild(name)))
	{
		wLog* root = WLog_GetRoot();
		if (!root)
			return NULL;
		if (!(log = WLog_New(name, root)))
			return NULL;
		if (!WLog_AddChild(root, log))
		{
			WLog_Free(log);
			return NULL;
		}
	}
	return log;
}

BOOL WLog_Init()
{
	return WLog_GetRoot() != NULL;
}

BOOL WLog_Uninit()
{
	DWORD index;
	wLog* child = NULL;
	wLog* root = g_RootLog;

	if (!root)
		return FALSE;

	for (index = 0; index < root->ChildrenCount; index++)
	{
		child = root->Children[index];
		WLog_Free(child);
	}

	WLog_Free(root);
	g_RootLog = NULL;
	return TRUE;
}
